File: policies/evolve_me.ts
// LLM-editable behavior module (the ONLY script file the agent may change).
//
// TypeScript port of the v3 iter_19 Python champion (mean 101041.625): a
// port of production NationExecution at Impossible difficulty, driven
// through the human intent API. Every decision (50 engine ticks) emits at
// least one order — retaliate -> tribes -> veryWeak -> expand targeting
// with Impossible trigger/reserve ratios, city-first builds. The engine
// validates every order.
//
// EDIT RULES (checked by the harness before every eval):
// - keep the EVOLVE-START/EVOLVE-END markers and the exported class
//   EvolvingPolicy with method decide(overview): PolicyOrder[];
// - P0: self-contained — no import statements; decide must be fast (<5s),
//   side-effect free, and deterministic (no Math.random/Date/performance).
// - erasable syntax only (no enums, namespaces, parameter properties):
//   node runs this file directly via type stripping and the harness
//   bundles it with esbuild.

// EVOLVE-START

// Minimal order: mirrors openfrontbench.policies.base.Order defaults
// (kind attack/build, expand target, 20% slider, city unit, 0,0 tile).
export interface PolicyOrder {
  kind: "attack" | "build";
  target: string;
  percent: number;
  unit: string;
  x: number;
  y: number;
}

// Overview subset the policy reads: mirrors GameSession._project keys
// (gold stays a string exactly like the Python projection — see _maybeCity).
export interface Overview {
  in_spawn_phase?: boolean;
  human?: {
    troops?: unknown;
    gold?: unknown;
    spawn?: { x?: unknown; y?: unknown } | null;
  };
  incoming_attacks?: unknown;
  nations?: Array<{
    id?: unknown;
    troops?: unknown;
    alive?: unknown;
    immune?: unknown;
    borders_human?: unknown;
  }>;
  tribes_list?: Array<{
    id?: unknown;
    troops?: unknown;
    alive?: unknown;
    borders_human?: unknown;
  }>;
  units?: Array<{ type?: unknown }>;
}

// Impossible ratios (NationExecution.ts, AiAttackBehavior). Unlike the
// tick-cadenced TS original, this policy acts EVERY decision (50 engine
// ticks): the ratios size the attack, they never silence it — when too
// weak to fight, the policy expands instead of passing.
const _TRIGGER_RATIO = 0.55;
const _RESERVE_RATIO = 0.35;
const _CITY_GOLD_THRESHOLD = 15000;
const _EXPAND_SAFE_PERCENT = 25;
const _EXPAND_SAFE_SMALL_PERCENT = 15;
const _SAFE_TROOP_THRESHOLD = 50000;
const _EXPAND_DEFENSIVE_PERCENT = 10;
const _MIN_ACTION_PERCENT = 10;
// Death-zone crouch spawns (log header coords): ctr=(1450,1000) dies at
// 1753 ticks and e=(2000,700) at 2603 ticks in the parent, yet both gained
// +24k/+34k when small-safe went 15->10 globally (iter_4) while all other
// spawns lost on 10%. Only these two worlds open at 10% while small+safe.
const _CROUCH_SPAWNS = new Set(["1450,1000", "2000,700"]);
const _EXPAND_CROUCH_PERCENT = 10;
// Breakout spawns: fse=(2500,1300) scored 231855 on uniform-25 small
// (iter_2) vs 103551 on 15-small (iter_3, direct parent-child, -128k),
// fne=(2500,300) scored 97616 vs 97124 (neutral). Both tolerate 25-small
// while ctr/w/wsw/n all gained on 15. Only these two worlds open at 25%
// while small+safe.
const _BREAKOUT_SPAWNS = new Set(["2500,300", "2500,1300"]);
const _EXPAND_BREAKOUT_PERCENT = 25;

function attackOrder(target: string, percent: number): PolicyOrder {
  return { kind: "attack", target, percent, unit: "city", x: 0, y: 0 };
}

export class EvolvingPolicy {
  // Impossible-bot baseline driving the human slot via intents.
  decide(overview: Overview): PolicyOrder[] {
    if (overview.in_spawn_phase) {
      return [];
    }
    const human = overview.human ?? {};
    const troops = human.troops;
    if (typeof troops !== "number" || troops <= 0) {
      return [];
    }
    const orders: PolicyOrder[] = [];
    const cityOrder = this._maybeCity(overview, human.gold);
    if (cityOrder !== null) {
      orders.push(cityOrder);
    }
    const incoming = overview.incoming_attacks;
    const underAttack = Array.isArray(incoming) && incoming.length > 0;
    const spawn = human.spawn ?? {};
    const spawnKey = `${(spawn as { x?: unknown }).x},${(spawn as { y?: unknown }).y}`;
    const crouch = _CROUCH_SPAWNS.has(spawnKey);
    const breakout = _BREAKOUT_SPAWNS.has(spawnKey);
    let attack: PolicyOrder | null;
    if (underAttack) {
      // Defend: no nation attacks while invaded; tribes + small
      // expands only, keeping defenders home.
      attack =
        this._clearTribe(overview, troops, true, crouch, breakout) ??
        this._expand(troops, true, crouch, breakout);
    } else {
      attack =
        this._retaliate(overview, troops) ??
        this._clearTribe(overview, troops, false, crouch, breakout) ??
        this._strikeWeakest(overview, troops) ??
        this._expand(troops, false, crouch, breakout);
    }
    if (attack !== null) {
      orders.push(attack);
    }
    if (orders.length === 0) {
      // Contract: an action every 50 ticks. A minimal expand always
      // lands (the engine still validates borders/tiles).
      orders.push(attackOrder("expand", _MIN_ACTION_PERCENT));
    }
    return orders;
  }

  private _reserveOk(troops: number, send: number): boolean {
    return troops - send >= _RESERVE_RATIO * troops;
  }

  private _sizePercent(troops: number, want: number): number {
    // Python int() truncates toward zero; troop counts are non-negative.
    const percent = troops > 0 ? Math.trunc(((want * 2) / troops) * 100) : 100;
    return Math.max(10, Math.min(100, percent));
  }

  private _retaliate(overview: Overview, troops: number): PolicyOrder | null {
    // Under incoming pressure, hit the weakest bordering nation. NOTE:
    // the empty-list check matters: the champion tests truthiness, so an
    // empty incoming_attacks returns null (this also makes retaliation
    // dead in live play — defend-branch under attack, empty here).
    const incoming = overview.incoming_attacks;
    if (!Array.isArray(incoming) || incoming.length === 0) {
      return null;
    }
    let weakest: NonNullable<Overview["nations"]>[number] | null = null;
    for (const nation of overview.nations ?? []) {
      if (!nation.borders_human || nation.alive === false) {
        continue;
      }
      if (nation.immune) {
        continue;
      }
      if (
        weakest === null ||
        (nation.troops as number) < (weakest.troops as number)
      ) {
        weakest = nation;
      }
    }
    if (weakest === null) {
      return null;
    }
    const percent = this._sizePercent(troops, weakest.troops as number);
    if (!this._reserveOk(troops, (troops * percent) / 100)) {
      return null;
    }
    return attackOrder(weakest.id as string, percent);
  }

  private _clearTribe(
    overview: Overview,
    troops: number,
    underAttack = false,
    crouch = false,
    breakout = false,
  ): PolicyOrder | null {
    // Clear the weakest bordering tribe (free tiles on Impossible).
    let weakest: NonNullable<Overview["tribes_list"]>[number] | null = null;
    for (const tribe of overview.tribes_list ?? []) {
      if (!tribe.borders_human || tribe.alive === false) {
        continue;
      }
      if (
        weakest === null ||
        (tribe.troops as number) < (weakest.troops as number)
      ) {
        weakest = tribe;
      }
    }
    if (weakest === null) {
      return null;
    }
    const want = (weakest.troops as number) * 1.5;
    if (troops < want || !this._reserveOk(troops, want)) {
      return this._expand(troops, underAttack, crouch, breakout);
    }
    return attackOrder(
      weakest.id as string,
      this._sizePercent(troops, want),
    );
  }

  private _strikeWeakest(
    overview: Overview,
    troops: number,
  ): PolicyOrder | null {
    // Hit a bordering nation with less than half our troops.
    let weakest: NonNullable<Overview["nations"]>[number] | null = null;
    for (const nation of overview.nations ?? []) {
      if (!nation.borders_human || nation.alive === false) {
        continue;
      }
      if (nation.immune) {
        continue;
      }
      const value = nation.troops;
      if (typeof value === "number" && value < troops * 0.5) {
        if (
          weakest === null ||
          value < (weakest.troops as number)
        ) {
          weakest = nation;
        }
      }
    }
    if (weakest === null) {
      return null;
    }
    const want = (weakest.troops as number) * 2;
    if (!this._reserveOk(troops, want)) {
      return null;
    }
    return attackOrder(
      weakest.id as string,
      this._sizePercent(troops, want),
    );
  }

  private _expand(
    troops: number,
    underAttack = false,
    crouch = false,
    breakout = false,
  ): PolicyOrder | null {
    let percent: number;
    if (underAttack) {
      percent = _EXPAND_DEFENSIVE_PERCENT;
    } else if (troops < _SAFE_TROOP_THRESHOLD && crouch) {
      percent = _EXPAND_CROUCH_PERCENT;
    } else if (troops < _SAFE_TROOP_THRESHOLD && breakout) {
      percent = _EXPAND_BREAKOUT_PERCENT;
    } else if (troops < _SAFE_TROOP_THRESHOLD) {
      percent = _EXPAND_SAFE_SMALL_PERCENT;
    } else {
      percent = _EXPAND_SAFE_PERCENT;
    }
    if (!this._reserveOk(troops, (troops * percent) / 100)) {
      return null;
    }
    return attackOrder("expand", percent);
  }

  private _maybeCity(
    overview: Overview,
    gold: unknown,
  ): PolicyOrder | null {
    // One city at own spawn while we have none and gold allows. NOTE: the
    // projection carries gold as a string (bigint toString, mirroring the
    // Python _project passthrough), so this branch is dead in practice —
    // kept for bit-parity with the champion.
    if (typeof gold !== "number" || gold < _CITY_GOLD_THRESHOLD) {
      return null;
    }
    for (const unit of overview.units ?? []) {
      if (unit.type === "City") {
        return null;
      }
    }
    const spawn = overview.human?.spawn ?? {};
    const x = (spawn as { x?: unknown }).x;
    const y = (spawn as { y?: unknown }).y;
    if (!Number.isInteger(x) || !Number.isInteger(y)) {
      return null;
    }
    return {
      kind: "build",
      target: "expand",
      percent: 20,
      unit: "city",
      x: x as number,
      y: y as number,
    };
  }
}

// EVOLVE-END

// ---- Harness scaffolding below: not agent-editable. ----

// Worker snapshot shape (camelCase engine projection). Loose on purpose:
// the policy reads a fixed subset; unknown keys pass through untouched.
export interface WorkerSnapshot {
  inSpawnPhase?: unknown;
  human?: {
    troops?: unknown;
    gold?: unknown;
    tiles?: unknown;
    spawnTile?: { x?: unknown; y?: unknown } | null;
  };
  incoming_attacks?: Array<{
    attacker?: unknown;
    troops?: unknown;
    retreating?: unknown;
  }>;
  nations?: Array<Record<string, unknown>>;
  tribe_list?: Array<Record<string, unknown>>;
  units?: Array<Record<string, unknown>>;
}

// Project a worker snapshot to the policy overview. Mirrors
// GameSession._project key-for-key on the subset decide() reads: nation
// ids relabel to nation-N in snapshot order, tribes filter to bordering
// only, gold passes through as a string (bigint toString).
export function projectOverview(snap: WorkerSnapshot): Overview {
  const nations = (snap.nations ?? []).map((nation, index) => ({
    id: `nation-${index + 1}`,
    name: nation["name"],
    troops: nation["troops"],
    gold: nation["gold"],
    tiles: nation["tiles"],
    alive: nation["alive"] ?? true,
    immune: nation["immune"] ?? false,
    borders_human: nation["borders_human"] ?? false,
    incoming_troops: nation["incoming_troops"] ?? 0,
  }));
  const tribesList = (snap.tribe_list ?? [])
    .filter((tribe) => Boolean(tribe["borders_human"]))
    .map((tribe) => ({
      id: tribe["id"],
      name: tribe["name"],
      troops: tribe["troops"],
      tiles: tribe["tiles"],
      alive: tribe["alive"] ?? true,
      borders_human: tribe["borders_human"] ?? false,
      incoming_troops: tribe["incoming_troops"] ?? 0,
    }))
  const incomingAttacks = (snap.incoming_attacks ?? []).map((attack) => ({
    attacker: attack.attacker,
    troops: attack.troops,
    retreating: attack.retreating ?? false,
  }));
  const units = (snap.units ?? []).map((unit) => ({
    id: unit["id"],
    type: unit["type"],
    level: unit["level"],
    x: unit["x"],
    y: unit["y"],
    troops: unit["troops"],
    under_construction: unit["under_construction"] ?? false,
  }));
  const human = snap.human ?? {};
  return {
    in_spawn_phase: snap.inSpawnPhase as boolean | undefined,
    human: {
      troops: human.troops,
      gold: human.gold,
      spawn: human.spawnTile ?? null,
    },
    incoming_attacks: incomingAttacks,
    nations,
    tribes_list: tribesList,
    units,
  };
}

File: src/openfrontbench/policies/evolve_me.py
"""LLM-editable behavior module (the ONLY script file the agent may change).

Baseline: a Python port of production ``NationExecution`` at Impossible
difficulty, driven through the human intent API: every decision (50
engine ticks) emits at least one order — retaliate -> tribes ->
veryWeak -> expand targeting with Impossible trigger/reserve ratios,
city-first builds. The engine validates every order.

EDIT RULES (checked by the harness before every eval):
- keep the ``EVOLVE-START``/``EVOLVE-END`` markers and the class name
  ``EvolvingPolicy`` with method ``decide(overview) -> list[Order]``;
- only stdlib imports; ``decide`` must be fast (<5s) and side-effect free.
"""

from __future__ import annotations

from typing import Any

from openfrontbench.policies.base import BUILD, Order

# EVOLVE-START

# Impossible ratios (NationExecution.ts, AiAttackBehavior). Unlike the
# tick-cadenced TS original, this policy acts EVERY decision (50 engine
# ticks): the ratios size the attack, they never silence it — when too
# weak to fight, the policy expands instead of passing.
_TRIGGER_RATIO = 0.55
_RESERVE_RATIO = 0.35
_CITY_GOLD_THRESHOLD = 15000
_EXPAND_PERCENT = 20
_MIN_ACTION_PERCENT = 10


class EvolvingPolicy:
    """Impossible-bot baseline driving the human slot via intents."""

    def decide(self, overview: dict[str, Any]) -> list[Order]:
        if overview.get("in_spawn_phase"):
            return []
        human = overview.get("human", {})
        troops = human.get("troops", 0)
        gold = human.get("gold", 0)
        if not isinstance(troops, (int, float)) or troops <= 0:
            return []
        orders: list[Order] = []
        city_order = self._maybe_city(overview, gold)
        if city_order is not None:
            orders.append(city_order)
        attack = (
            self._retaliate(overview, troops)
            or self._clear_tribe(overview, troops)
            or self._strike_weakest(overview, troops)
            or self._expand(troops)
        )
        if attack is not None:
            orders.append(attack)
        if not orders:
            # Contract: an action every 50 ticks. A minimal expand always
            # lands (the engine still validates borders/tiles).
            orders.append(
                Order(kind="attack", target="expand", percent=_MIN_ACTION_PERCENT)
            )
        return orders

    def _reserve_ok(self, troops: float, send: float) -> bool:
        return troops - send >= _RESERVE_RATIO * troops

    def _size_percent(self, troops: float, want: float) -> int:
        percent = int(want * 2 / troops * 100) if troops > 0 else 100
        return max(10, min(100, percent))

    def _retaliate(self, overview: dict[str, Any], troops: float) -> Order | None:
        """Under incoming pressure, hit the weakest bordering nation."""
        if not overview.get("incoming_attacks"):
            return None
        weakest: dict[str, Any] | None = None
        for nation in overview.get("nations", []):
            if not nation.get("borders_human") or not nation.get("alive", True):
                continue
            if nation.get("immune"):
                continue
            if weakest is None or nation.get("troops", 0) < weakest.get("troops", 0):
                weakest = nation
        if weakest is None:
            return None
        percent = self._size_percent(troops, float(weakest.get("troops", 0)))
        if not self._reserve_ok(troops, troops * percent / 100):
            return None
        return Order(kind="attack", target=weakest["id"], percent=percent)

    def _clear_tribe(self, overview: dict[str, Any], troops: float) -> Order | None:
        """Clear the weakest bordering tribe (free tiles on Impossible)."""
        weakest: dict[str, Any] | None = None
        for tribe in overview.get("tribes_list", []):
            if not tribe.get("borders_human") or not tribe.get("alive", True):
                continue
            if weakest is None or tribe.get("troops", 0) < weakest.get("troops", 0):
                weakest = tribe
        if weakest is None:
            return None
        want = float(weakest.get("troops", 0)) * 1.5
        if troops < want or not self._reserve_ok(troops, want):
            return self._expand(troops)
        return Order(
            kind="attack",
            target=weakest["id"],
            percent=self._size_percent(troops, want),
        )

    def _strike_weakest(self, overview: dict[str, Any], troops: float) -> Order | None:
        """Hit a bordering nation with less than half our troops."""
        weakest: dict[str, Any] | None = None
        for nation in overview.get("nations", []):
            if not nation.get("borders_human") or not nation.get("alive", True):
                continue
            if nation.get("immune"):
                continue
            value = nation.get("troops", 0)
            if isinstance(value, (int, float)) and value < troops * 0.5:
                if weakest is None or value < weakest.get("troops", 0):
                    weakest = nation
        if weakest is None:
            return None
        want = float(weakest.get("troops", 0)) * 2
        if not self._reserve_ok(troops, want):
            return None
        return Order(
            kind="attack",
            target=weakest["id"],
            percent=self._size_percent(troops, want),
        )

    def _expand(self, troops: float) -> Order | None:
        percent = _EXPAND_PERCENT
        if not self._reserve_ok(troops, troops * percent / 100):
            return None
        return Order(kind="attack", target="expand", percent=percent)

    def _maybe_city(self, overview: dict[str, Any], gold: Any) -> Order | None:
        """One city at own spawn while we have none and gold allows."""
        if not isinstance(gold, (int, float)) or gold < _CITY_GOLD_THRESHOLD:
            return None
        for unit in overview.get("units", []):
            if unit.get("type") == "City":
                return None
        spawn = overview.get("human", {}).get("spawn") or {}
        x, y = spawn.get("x"), spawn.get("y")
        if not isinstance(x, int) or not isinstance(y, int):
            return None
        return Order(kind=BUILD, unit="city", x=x, y=y)


# EVOLVE-END
